File: 2020_kakao_blind_recruitment/string_compression.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int solution(string s) {
    int sLength = (int)s.length();
    int answer = sLength;
    
    for (int unit = 1; unit <= sLength; unit++) {
        string words = s.substr(0, unit);
        string str = "";
        
        // unit 단위로 자르기
        int wordsCount = 1;
        for (int i = unit; i < sLength; i+=unit) {
            // unit 단위로 자르기 때문에 for문을 통해 문자열이 일치하는지 체크한다.
            bool ok = true;
            for (int k = 0 ; k < words.length(); k++) {
                if (words[k] != s[i + k]) {
                    ok = false;
                    break;
                }
            }
            
            // 문자열이 일치하면 카운트 증가
            if (ok) wordsCount++;
            else {
                // 문자열이 일치하지 않으면 압축된 문자열 만들기
                // 주의할 점은 문자가 1개이면, 압축된 문자열에 포함시키지 않는다.
                if (wordsCount > 1) str += to_string(wordsCount);
                str += words;
                words = s.substr(i, unit);
                wordsCount = 1;
            }
        }
        
        // 마지막 남은 문자를 압축된 문자열에 더한다.
        if (wordsCount > 1) str += to_string(wordsCount);
        str += words;
        
        answer = min(answer, (int)str.length());
    }
    
    return answer;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    
    cout << solution("aabbaccc") << '\n';
    cout << solution("ababcdcdababcdcd") << '\n';
    cout << solution("abcabcdede") << '\n';
    cout << solution("abcabcabcabcdededededede") << '\n';
    cout << solution("xababcdcdababcdcd") << '\n';
    
    return 0;
}

File: 2020_kakao_blind_recruitment/string_compression.js
function solution(s) {
  if (s.length === 1) return 1;

  let answer = -1;

  const unitCount = parseInt(s.length / 2);

  for (let unit = 1; unit <= unitCount; unit++) {
    let tempString = "";

    let index = 0;
    while (index < s.length) {
      const char = s.slice(index, index + unit);

      index += unit;

      let charCount = 0;
      while (s.slice(index, index + unit) === char) {
        charCount++;
        index += unit;
      }

      tempString += !charCount ? `${char}` : `${charCount + 1}${char}`;
    }

    if (answer === -1 || answer > tempString.length) {
      answer = tempString.length;
    }
  }

  return answer;
}

console.log(solution("aabbaccc"));
console.log(solution("ababcdcdababcdcd"));
console.log(solution("abcabcdede"));
console.log(solution("abcabcabcabcdededededede"));
console.log(solution("xababcdcdababcdcd"));

// "aabbaccc"	7
// "ababcdcdababcdcd"	9
// "abcabcdede"	8
// "abcabcabcabcdededededede"	14
// "xababcdcdababcdcd"	17
